packages/main/src/Form.ts

```diff
@@ -0,0 +1,437 @@
+			return (itemB as FormGroup)?.children?.length - (itemA as FormGroup)?.children?.length;
+		});
+
+		sortedItems.forEach((item: FormGroup | FormItem, idx: number) => {
+			(item as FormGroup).colsXl = this.getGroupsColSpan(this.columnsXl, itemsCount, idx, (item as FormGroup));
+			(item as FormGroup).colsL = this.getGroupsColSpan(this.columnsL, itemsCount, idx, (item as FormGroup));
+			(item as FormGroup).colsM = this.getGroupsColSpan(this.columnsM, itemsCount, idx, (item as FormGroup));
+			(item as FormGroup).colsS = this.getGroupsColSpan(this.columnsS, itemsCount, idx, (item as FormGroup));
+		});
+	}
+
+	getGroupsColSpan(cols: number, groups: number, index: number, group: FormGroup): number {
+		// Case 0: column span is set from outside.
+		if (group.columnSpan) {
+			return group.columnSpan;
+		}
+
+		// CASE 1: The number of available columns match the number of groups, or only 1 column is available - each group takes 1 column.
+		// For example: 1 column - 1 group, 2 columns - 2 groups, 3 columns - 3 groups, 4columns - 4 groups
+		if (cols === 1 || cols <= groups) {
+			return 1;
+		}
+
+		// CASE 2: The number of available columns IS multiple of the number of groups.
+		// For example: 2 column - 1 group, 3 columns - 1 groups, 4 columns - 1 group, 4 columns - 2 groups
+		if (cols % groups === 0) {
+			return cols / groups;
+		}
+
+		// CASE 3: The number of available columns IS NOT multiple of the number of groups.
+		const MIN_COL_SPAN = 1;
+		const delta = cols - groups;
+
+		// 7 cols & 4 groups => 2, 2, 2, 1
+		if (delta <= groups) {
+			return index < delta ? MIN_COL_SPAN + 1 : MIN_COL_SPAN;
+		}
+
+		// 7 cols & 3 groups => 3, 2, 2
+		return index === 0 ? MIN_COL_SPAN + (delta - groups) + 1 : MIN_COL_SPAN + 1;
+	}
+
+	get hasGroupItems(): boolean {
+		return this.items.some((item: FormGroup | FormItem) => item.isGroup);
+	}
+
+	get hasCustomHeader(): boolean {
+		return !!this.header.length;
+	}
+
+	get ariaLabelledByID(): string | undefined {
+		return this.hasCustomHeader ? undefined : `${this._id}-header-text`;
+	}
+
+	get itemsInfo(): Array<ItemsInfo> {
+		return this.items.map((item: FormGroup | FormItem) => {
+			return {
+				item,
+				classes: `ui5-form-column-spanL-${(item as FormGroup).colsL} ui5-form-column-spanXL-${(item as FormGroup).colsXl} ui5-form-column-spanM-${(item as FormGroup).colsM} ui5-form-column-spanS-${(item as FormGroup).colsS}`,
+				items: Array.from((item as FormGroup).children) as Array<FormItem>,
+			};
+		});
+	}
+
+	createStepCSSStyleSheet() {
+		[
+			{ breakpoint: "S", columns: this.columnsS },
+			{ breakpoint: "M", columns: this.columnsM },
+			{ breakpoint: "L", columns: this.columnsL },
+			{ breakpoint: "XL", columns: this.columnsXl },
+		].forEach(step => {
+			const styleSheet = this.getStepCSSStyleSheet(step.breakpoint, step.columns);
+			if (styleSheet) {
+				this.shadowRoot!.adoptedStyleSheets.push(styleSheet);
```

Comment:
we still support other stylesheet solutions like inline styles, links. Is this save to be used?

Reply:
No, in 2.0 - we will only use adoptedStyleSheets